File: LINKLIST/JAVA/doublyCLL.java
class node
{
    public int data ;
    public node next;
    public node prev;
}

class doublyCLLX
{
    private node first;
    private node last ;
    private int icount;

    public
    doublyCLLX()
    {
        first = null;
        last = null;
        icount = 0;
    }

    void insertfirst(int no)
    {
        node newn = new node();

        newn.data = no;
        newn.next = null;
        newn.prev = null;

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first.prev = newn;

            first = newn;   
        }
        last.next= first;
        first.prev = last;
        icount++;

    }

    void insertlast(int no)
    {
        node newn = new node();

        newn.data = no;
        newn.next = null;
        newn.prev = null;

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            last.next = newn;
            newn.prev = last;

            last = newn;
        }
        last.next = first;
        first.prev = last;
        icount++;

    }

    void insertatpos(int no,int pos)
    {
        int icnt = 0;
        node newn = new node();
        node temp = new node();

        if(pos < 1 || pos > icount + 1) 
        {
            System.out.println("invalid position");
            return;
        }

        if(pos == 1)
        {
            insertfirst(no);
        }
        else if( pos == icount + 1)
        {
            insertlast(no);
        }
        else
        {
            temp = first;

            newn.data = no;
            newn.next = null;
            newn.prev = null;

            for(icnt = 1;icnt < pos - 1;icnt++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next.prev = newn;

            temp.next  = newn;
            newn.prev = temp;

            icount++;
        }
    }

    void deletefirst()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if (first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            first = first.next;

        }
        last.next = first;
        first.prev= last;

        icount--;
    }

    void deletelast()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if (first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            last = last.prev;
        }
        last.next = first;
        first.prev = last;

        icount--;

    }

    void deleteatpos(int pos)
    {
        if(pos < 1 || pos > icount) 
        {
            System.out.println("invalid position");
            return;
        }

        if(pos == 1)
        {
            deletefirst();
        }
        else if( pos == icount)
        {
            deletelast();
        }
        else
        {
            node temp = new node();
            int icnt  = 0;
            temp = first;

            for(icnt = 1;icnt < pos-1;icnt++)
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;
            temp.next.prev = temp;

            icount--;
        }
    }

    void display()
    {
        node temp = new node();

        temp = first;

        if(first == null && last == null)
        {
            return;
        }

        System.out.print(" <=>");
        do
        {
            System.out.print(" | "+ temp.data +" | <=> ");
            temp = temp.next;
        }while(temp != last.next);
        System.out.println();
        
    }

    int count ()
    {
        return icount;
    }

}

class doublyCLL
{
    public static void main(String a[])
    {
        doublyCLLX sobj = new doublyCLLX();
        int iret = 0;

        sobj.insertfirst(51);
        sobj.insertfirst(21);
        sobj.insertfirst(11);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.insertlast(101);
        sobj.insertlast(111);
        sobj.insertlast(121);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.insertatpos(105,5);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);


        sobj.deletefirst();

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.deletelast();

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.deleteatpos(5);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

    }
}

File: LINKLIST/C/doublyCLL.c
 a#include <stdio.h>
#include<stdlib.h>

struct node
{
    int data ;
    struct node * next;
    struct node * prev;
};

typedef struct node  NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

void insertfirst(PPNODE first,PPNODE last,int no)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn ->data = no;
    newn->next = NULL;
    newn->prev = NULL;

    if(*first == NULL && *last == NULL)
    {
        *first = newn;
        *last = newn;

        (*last)->next = *first;
        (*first)-> prev = *last;
    }
    else
    {
        newn->next = *first;
        (*first)->prev = newn;
        *first = newn;

        (*last)->next = *first;
        (*first)-> prev = *last;

    }

}

void insertlast(PPNODE first,PPNODE last,int no)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn ->data = no;
    newn->next = NULL;
    newn->prev = NULL;

    if(*first == NULL && *last == NULL)
    {
        *first = newn;
        *last = newn;

        (*last)->next = *first;
        (*first)-> prev = *last;
    }
    else
    {
        (*last)->next = newn;
        newn-> prev = *last;
        *last = newn;
        
        (*last)->next = *first;
        (*first)-> prev = *last;

    }
}

void deletefirst(PPNODE first,PPNODE last)
{
    PNODE temp = NULL ;

    if(*first == NULL && *last == NULL)
    {
        return;
    }
    else if(*first == *last)
    {
        free(*first);
        *first = NULL;
        *last = NULL;
    }
    else
    {
        temp = *first;
        *first = temp -> next;
        free(temp);

        (*last)-> next = *first;
        (*first)-> prev = *last;

    }
}

void deletelast(PPNODE first,PPNODE last)
{
    PNODE temp = NULL;

    if(*first == NULL && *last == NULL)
    {
        return;
    }
    else if(*first == *last)
    {
        free(*first);
        *first = NULL;
        *last = NULL;
    }
    else
    {
        temp = (*last)->prev;
        free(*last);
        *last = temp;

        (*last)->next = *first;
        (*first)->prev = *last;
    }
}

void display(PNODE first,PNODE last)
{
    PNODE temp = NULL;
    temp = first;

    if(first == NULL && last == NULL)
    {
        return;
    }

    printf("NULL");
    do
    {
        printf("| %d | <=> ",temp -> data);
        temp = temp -> next;
    }while(temp != first);
    printf("NULL\n");

}

int count(PNODE first,PNODE last)
{
    PNODE temp = NULL;
    int icount = 0;
    temp = first;

    if(first == NULL && last == NULL)
    {
        return icount;
    }

    do
    {
        icount++;
        temp = temp -> next;
    }while(temp != first);

    return icount;
  
}

void insertatpos(PPNODE first,PPNODE last,int no,int pos)
{
    PNODE temp = NULL;
    PNODE newn = NULL;
    int icnt = 0;
    int icount = 0;

    icount = count(*first,*last);

    if(pos < 1 || pos > icount + 1)
    {
        printf("invalid position\n");
        return;
    }

    if(pos == 1)
    {
        insertfirst(first,last,no);
    }
    else if(pos == icount + 1)
    {
        insertlast(first,last,no);
    }
    else
    {
        temp = *first;

        newn = (PNODE)malloc(sizeof (NODE));

        newn -> data = no;
        newn -> next = NULL;
        newn -> prev = NULL;

        for(icnt = 1;icnt < pos-1; icnt++)
        {
            temp = temp ->next;
        }
        newn->next = temp ->next;
        temp ->next ->prev = newn;
        temp->next = newn;
        newn->prev = temp;

    }
}
void deleteatpos(PPNODE first,PPNODE last,int pos)
{
    PNODE target = NULL;
    PNODE temp = NULL;
    int icnt = 0;
    int icount = 0;

    icount = count(*first,*last);

    if(pos < 1 || pos > icount)
    {
        printf("invalid position\n");
        return;
    }

    if(pos == 1)
    {
        deletefirst(first,last);

    }
    else if(pos == icount)
    {
        deletelast(first,last);
    }
    else
    {
        temp = *first;

        for(icnt = 1;icnt < pos -1;icnt++)
        {
            temp = temp -> next;
        }
        target = temp -> next;
        temp -> next = target ->next;
        temp -> next ->prev = temp;

        free(target);

    }
}

int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;
    int iret = 0;

    insertfirst(&head,&tail,51);
    insertfirst(&head,&tail,21);
    insertfirst(&head,&tail,11);

    display(head,tail);
    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    insertlast(&head,&tail,101);
    insertlast(&head,&tail,111);
    insertlast(&head,&tail,121);

    display(head,tail);
    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    deletefirst(&head,&tail);

    display(head,tail);
    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    deletelast(&head,&tail);

    display(head,tail);
    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    insertatpos(&head,&tail,105,5);

    display(head,tail);
    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    deleteatpos(&head,&tail,5);

    display(head,tail);
    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    
    return 0;
}
